Add package & NixOS module for journaldriver

tazjin opened the pull request:
###### Motivation for this change

Add support for the [journaldriver](https://github.com/aprilabank/journaldriver) log forwarding agent, primarily for use on Google Cloud Platform instances.

This PR adds a derivation for building `journaldriver`, as well as a NixOS module that configures a user and service for running the log forwarder.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

We have deployed this module on a bunch of machines in our test environments (both on Google Cloud machines and VMs hosted elsewhere) and it works as expected.

tazjin committed:
journaldriver: init at 1.0.0

tazjin committed:
nixos/journaldriver: add module for journaldriver log forwarder

Adds a module for running the journaldriver log forwarding agent via
systemd.

The agent can be deployed on both GCP instances and machines hosted
elsewhere to forward all logs from journald to Stackdriver Logging.

Consult the module options and upstream documentation for more
information.

Implementation notes:

* The service unit is configured to use systemd's dynamic user feature
  which will let systemd set up the state directory and appropriate
  user configuration at unit launch time instead of hardcoding it.

* The module depends on `network-online.target` to prevent a situation
  where journaldriver is failing and restarting multiple times before
  the network is online.

Infinisil commented on pkgs/tools/misc/journaldriver/default.nix:
```diff
@@ -0,0 +1,27 @@
+{ lib, fetchFromGitHub, rustPlatform, pkgconfig, openssl, systemd }:
+
+rustPlatform.buildRustPackage rec {
+  name        = "journaldriver";
```
The version should be appended to the name, separated by a dash

tazjin replied:
Fixed!

Infinisil commented on pkgs/tools/misc/journaldriver/default.nix:
```diff
@@ -0,0 +1,27 @@
+  };
+
+  # Disable tests in nixpkgs
+  doCheck = false;
```
What's the reason for the tests being disabled?

tazjin replied:
Mostly decided to do that due to the time it takes to build the entire dependency chain, which will happen twice for running tests & non-tests.

The release versions have already had tests run (otherwise they wouldn't have gone into upstream master) and since all dependencies are pinned it should be fine.

Normally I'd package a Rust tool by using the `buildRustCrate` (which doesn't run tests at all), but that's not resolving the dependency on libsystemd correctly for some reason and has no obvious (read: documented) way to override the dependencies it figures out.

If you think tests should be enabled I don't mind enabling them :)

Infinisil replied:
As long as the tests work, I think they should be on by default. Hydra will have it in the cache anyways, so it doesn't really matter how long it takes for the end user. Usually the only good reason to disable tests is because the tests require network access or something else that Nix doesn't have in the sandbox.

tazjin replied:
:+1: Updating with tests enabled.

Infinisil commented on pkgs/tools/misc/journaldriver/default.nix:
```diff
@@ -0,0 +1,27 @@
+  # Disable tests in nixpkgs
+  doCheck = false;
+
+  buildInputs = [ pkgconfig openssl systemd.dev ];
```
`buildInputs` should choose the dev outputs automatically, so I think it should also work with `s/systemd.dev/systemd`

Infinisil replied:
Also, `pkgconfig` should go in `nativeBuildInputs`

Infinisil commented on nixos/modules/services/logging/journaldriver.nix:
```diff
@@ -0,0 +1,117 @@
+
+  config = mkIf cfg.enable {
+    users.extraUsers.journaldriver = {
+      isSystemUser = true;
```
Why is this set to `true`?

tazjin replied:
As far as I understand the differentiation this is only to indicate that the user account is not used by a human, which results in the UID of the account being allocated in a different range.

Infinisil replied:
Hmm, I'm not entirely sure, there are a couple services that use this, but others set a fixed id in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/ids.nix#L36. I hope somebody else can comment on this

tazjin replied:
The comment at the top of `ids.nix` sounds to me like the preferred way is to not hardcode an ID unless necessary.

It may be possible to use the `DynamicUser=` option in systemd. Journaldriver needs to have a persistent file to store its cursor position in, this should be possible to achieve with `StateDirectory=` in systemd - it's not obvious how to grant the dynamic user membership in the journal access group though.

tazjin replied:
Update:

This configuration seems to work without specifying a user at all:

```
      serviceConfig = {
        Restart        = "always";
        DynamicUser    = true;
        StateDirectory = "journaldriver";

        # This group is required for accessing journald.
        SupplementaryGroups = "systemd-journal";
      };
```

Infinisil commented on nixos/modules/services/logging/journaldriver.nix:
```diff
@@ -0,0 +1,117 @@
+    };
+
+    systemd.services.journaldriver = {
+      inherit environment;
```
You can move the section at the start here, and change it to something like this:

```nix
environment = {
  RUST_LOG = cfg.logLevel;
  LOG_NAME = cfg.logName;
  # ...
}
```

This works because `environment` allows `null` values which has the effect of not setting the vars at all already.

Infinisil commented on nixos/modules/services/logging/journaldriver.nix:
```diff
@@ -0,0 +1,117 @@
+
+    logLevel = mkOption {
+      type = types.string;
+      default = "info";
```
What valid values are there? You can change the type to `types.enum [ "info" "debug" ... ]` to only allow certain values.

tazjin replied:
The values aren't constrained by an enum in this case, Rust's logging library accepts pretty complex configuration via this environment variable (for example one could enable trace logging for the HTTP client used within while not touching the log-level of the rest of the service).

infinisil reviewed (changes requested):
How can one know whether its a GCP machine or not?

And a nitpick: Only half of the option declarations have the `=` lined up nicely :)

tazjin commented:
@Infinisil Thanks for the review! Good feedback :ok_hand: 

> How can one know whether its a GCP machine or not?

Hmm, presumably people know who their hosting provider is ... I'd hope? Technically one could try to run some kind of detection by attempting to speak to Google's instance metadata service, but that should probably be a change in journaldriver and not something that Nix should do.

I'm pushing an updated version that addresses some of your points and will quickly test it on some test machines.

tazjin commented:
Update: Tests successful :)

Infinisil commented on nixos/modules/services/logging/journaldriver.nix:
```diff
@@ -0,0 +1,112 @@
+        "LOG_NAME"                       = cfg.logName;
+        "LOG_STREAM"                     = cfg.logStream;
+        "GOOGLE_CLOUD_PROJECT"           = cfg.googleCloudProject;
+        "GOOGLE_APPLICATION_CREDENTIALS" = cfg.applicationCredentials;
```
I think it would be nicer to not use the `"` here, `RUST_LOG = cfg.logLevel` works just fine.

tazjin replied:
:+1: Updated

Infinisil commented on nixos/modules/services/logging/journaldriver.nix:
```diff
@@ -0,0 +1,112 @@
+    };
+
+    logLevel = mkOption {
+      type        = types.string;
```
Almost missed this: The type should be `types.str` instead: https://github.com/NixOS/nixpkgs/blob/7296c261f782ae525bdaf2a8c885e0d2c0f640e1/lib/types.nix#L203-L205

The same for all other options that use `types.string` except `applicationCredentials`, which should be changed to a path: `types.path`.

tazjin replied:
Fixed!

infinisil reviewed (changes requested):
Nice! Another thing I'm unsure about is `network.target` vs `network-online.target`. I'd think that most services would require `network-online.target`, but there are vastly more services actually using `network.target`. I suspect that's just because of confusion regarding those two, `network-online.target` seems to be the right one here.

One last thing to change and then this is looking good to me :)

tazjin commented:
Regarding the network targets: I read up on that a bit before making the change and according to [this article](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/), the `-online` target should be the correct one.

Infinisil commented on nixos/modules/services/logging/journaldriver.nix:
```diff
@@ -0,0 +1,112 @@
+
+        # The directory is set up and permissions are configured by
+        # systemd via the `StateDirectory` configuration option.
+        CURSOR_POSITION_FILE = "/var/lib/private/journaldriver/cursor.pos";
```
Reading `man systemd.exec` on `StateDirectory` tells me that symlinks should be set up to point `/var/lib/journaldriver -> /var/lib/private/journaldriver`, so I think the `private/` infix could be dropped.

Infinisil commented:
Other than that, I'd squash the 3 module commits into 1, and then I got nothing more to say :)

tazjin commented:
@Infinisil Upstream now uses `/var/lib/journaldriver/` as the default folder and I've updated the sources, so the cursor position setting is removed from here completely.

I've squashed the commits together now. Thanks for your review, I learned a bunch of interesting things that aren't immediately obvious from the Nix docs :)

tazjin commented:
Verified the final version in our test environment, good to go from my side!

infinisil reviewed (approved):
Looks good to me, nice!

fpletz reviewed (approved):
Awesome! Thanks for the PR @tazjin and @Infinisil for the thorough review. :beers:

tazjin commented:
@fpletz Thank you! :)
